Created a class and screen for JJ Slits

rsmm97 opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
## Description
<!--- Describe your changes in detail -->
Added a class for JJSlits_AT_C8_HV within slits.py.  Along with this was the creation of a typhos controls screen.
## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
This is a necessary addition since the new JJ Slits model: AT-C8-HV move as a pair compared to the regular slits which can move individually. With the creation of this class and screen it can easily be used with various JJSlits_AT_C8_HV using happi and typhos.
## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
This has been tested within the Test rack in building 901. Tested to make sure each axis of slits moves correctly and corresponds to controls screen.
## Where Has This Been Documented?
<!--  Include where the changes made have been documented. -->
<!--  This can simply be  a comment in the code or updating a docstring -->
There is a comment within the new class describing the parameters and use.
<!--
## Screenshots (if appropriate):
-->
JJSlits have been moved and were not connected when screenshots were taken, hence the red boxes/errors.

<img width="482" alt="jjslits_ipython1" src="https://user-images.githubusercontent.com/88116804/165127518-c5893bcc-7ee0-4210-8928-bae5d2002e6e.PNG">

<img width="917" alt="jjslits_screen1" src="https://user-images.githubusercontent.com/88116804/165125626-4e081b33-320e-4f97-86e1-4e2637cf2e8f.PNG">

<img width="919" alt="jjslits_screen2" src="https://user-images.githubusercontent.com/88116804/165125644-073eb269-e628-4186-954f-72a6adacc2fd.PNG">


## Pre-merge checklist
- [x] Code works interactively
- [ ] Code contains descriptive docstrings, including context and API
- [ ] New/changed functions and methods are covered in the test suite where possible
- [ ] Test suite passes locally
- [ ] Test suite passes on travis
- [ ] Ran docs/pre-release-notes.sh and created a pre-release documentation page
- [ ] Pre-release docs include context, functional descriptions, and contributors as appropriate

klauer commented on pcdsdevices/slits.py:
```diff
@@ -643,3 +643,18 @@ class SimLusiSlits(LusiSlits):
     ywidth = Cpt(FastMotor)
     xcenter = Cpt(FastMotor)
     ycenter = Cpt(FastMotor)
+
```
There's a style issue here (two blank lines between class definitions) that will cause continuous integration checks of your pull request to fail.

Did you configure pre-commit? It would have caught this issue before even letting you commit the changes. Usage is documented briefly here:
https://confluence.slac.stanford.edu/display/PCDS/How+to+add+a+device+to+pcdsdevices#Howtoaddadevicetopcdsdevices-Pre-commit

Alternatively, see the `flake8` section above that.

klauer commented on pcdsdevices/slits.py:
```diff
@@ -643,3 +643,18 @@ class SimLusiSlits(LusiSlits):
+    Parameters
+    ---------
+    prefix : str
+    Base PV for the JJ X-RAY PLC System
```
```suggestion
        Base PV for the JJ X-RAY PLC System
```

klauer commented on pcdsdevices/slits.py:
```diff
@@ -643,3 +643,18 @@ class SimLusiSlits(LusiSlits):
     ycenter = Cpt(FastMotor)
+
+class JJSlits(SlitsBase):
+    '''
```
Please add a descriptive string for the device:
* I believe JJ x-ray offers a variety of models of slits. Which does this specifically apply to?
* Is the class name as specific as it should be to disambiguate it from other slits on the beamline, and future JJ x-ray slits?

Mbosum replied:
@klauer this class is specifically for the AT-C8-HV model which have combined motors for a center motion, and a gap motion, as opposed to the other slits which have individual motors for each of the blades. Do you have any suggestions for a new class name?

klauer replied:
I think what you changed it to is reasonable. We're in the middle of trying to come up with a naming guide - like our style guide - so that there should be one unambiguous name to choose for a device.

Two things, though:
* This is still missing a docstring that includes the model number and relevant information
* The display filename will need to be renamed accordingly

klauer reviewed (changes requested):
There are some style and docstring changes needed here.

Please also add a pre-release changelog.

ZLLentz commented on docs/source/upcoming_release_notes/1001-jj-slits.rst:
```diff
@@ -0,0 +1,30 @@
+
+New Devices
+-----------
+- JJSlits class within slits.py
```
Probably needs more text here, just like in the empty class docstring: what exactly is this for? The new beckhoff controlled slits?

ZLLentz commented on docs/source/upcoming_release_notes/1001-jj-slits.rst:
```diff
@@ -0,0 +1,30 @@
+
+Contributors
+------------
+- Ricardo Martinez (rsmm97)
```
```suggestion
- rsmm97
```

ZLLentz commented on pcdsdevices/slits.py:
```diff
@@ -643,3 +643,18 @@ class SimLusiSlits(LusiSlits):
+    ---------
+    prefix : str
+    Base PV for the JJ X-RAY PLC System
+    name : str
```
```suggestion
    name : str, keyword-only
```

rsmm97 commented:
I have changed the class name to 'JJSlits_AT_C8_HV' which corresponds to the model number and added the pre-release document. I am not sure how to resubmit for a second pull-request check.

ZLLentz commented:
You missed the most important suggestion: a description in the class docstring and in the release notes as to what the device is.

rsmm97 commented:
I added some more info on the release notes doc.

ZLLentz commented:
Hey, do we need to have a quick meeting about this? I think we aren't communicating very well here.

The class name and model number are not very important. Absolutely a minor detail. There was actually not much of an explicit need to rename it- we can figure that out later and keep things backwards-compatible.

What is important is to answer the most basic user questions, in order of importance:
- What is a JJSlit?
- When do I use this class?
- How do I use this class?

Someone who reads the file should be able to understand what the class is for, and someone who owns the physical device should be able to figure out which class is the one that matches.

Right now, I can't glean any of that from the device docstring, nor can I from the release notes.

You know very much exactly what this device is because you are working on it. I have some insider knowledge about it as well. But we're not the only people who need to use this. It isn't even mentioned anywhere that this only works for the Beckhoff controls version of the JJSlits.

The documentation has none of the necessary context to be used by people in the future.

Here are some examples of docstrings that have context:
https://github.com/pcdshub/pcdsdevices/blob/e69e33cdde0347a6b05999276aa643353125434f/pcdsdevices/valve.py#L171-L180
https://github.com/pcdshub/pcdsdevices/blob/e69e33cdde0347a6b05999276aa643353125434f/pcdsdevices/mirror.py#L561-L573
https://github.com/pcdshub/pcdsdevices/blob/e69e33cdde0347a6b05999276aa643353125434f/pcdsdevices/attenuator.py#L1036-L1053

rsmm97 commented:
I updated the docstring and release document. Let me know if it is missing something.

ZLLentz commented:
The new docstring is very good and will hopefully be helpful in the screen too!
I want to merge this, however the automated checks are failing on things like import sorting and trailing whitespace. I'm going to leave the output from the checks here: https://app.travis-ci.com/github/pcdshub/pcdsdevices/jobs/568468088

I'm going to make suggestions that will:
- satisfy the automated checks
- conform the pre-release notes to the instructions at https://github.com/pcdshub/pcdsdevices/blob/master/docs/source/upcoming_release_notes/template-full.rst

If github lets me, I will forcibly apply these suggestions now and merge the PR before stopping work for the day. If github does not let me, I'll leave the suggestions for tomorrow.

ZLLentz commented on docs/source/upcoming_release_notes/1001-jj-slits.rst:
```diff
@@ -0,0 +1,30 @@
+-----------
+- N/A
+
+Features
+--------
+- New class and typhos screen for controlling JJSlits class.
+
+Device Updates
+--------------
+- Added JJSlits class within slits.py and a corresponding typhos screen.
+
+New Devices
+-----------
+- Model: JJ X-Ray AT-C8-HV
```
```suggestion
Features
--------
- N/A

Device Updates
--------------
- N/A

New Devices
-----------
- New `JJSlits` class and typhos screen for controlling JJSlits model AT-C8-HV with Beckhoff controls.
```

ZLLentz commented on pcdsdevices/slits.py:
```diff
@@ -28,7 +28,7 @@
 from .device import GroupDevice
 from .device import UpdateComponent as UpCpt
-from .epics_motor import BeckhoffAxisNoOffset
+from .epics_motor import BeckhoffAxisNoOffset, BeckhoffAxis
```
```suggestion
from .epics_motor import BeckhoffAxis, BeckhoffAxisNoOffset
```

ZLLentz commented on pcdsdevices/slits.py:
```diff
@@ -643,3 +643,33 @@ class SimLusiSlits(LusiSlits):
+class JJSlits(SlitsBase):
+    '''
+    Beckhoff Controlled AT-C8-HV JJ X-Ray Slits.
+    
```
```suggestion

```

ZLLentz commented on pcdsdevices/slits.py:
```diff
@@ -643,3 +643,33 @@ class SimLusiSlits(LusiSlits):
+    the position 'center' and opening 'width' of the aperture. The 'JJSlits'
+    class instantiates four sub-devices: 'JJSlits.xwidth', 'JJSlits.xcenter',
+    'JJSlits.ywidth', 'JJSlits.ycenter'.
+    
```
```suggestion

```

ZLLentz reviewed (approved):
Merging on successful automated checks

ZLLentz commented on pcdsdevices/slits.py:
```diff
@@ -643,3 +643,31 @@ class SimLusiSlits(LusiSlits):
+    The JJSlits class defines motor components for the specific JJ X-Ray slits
+    model AT-C8-HV. This model of JJSlits holds four in-vacuum motors that set
+    the position and opening of the aperture in two dimensions. For each
+    dimension, one motor defines the aperture opening, and another determines the
+    position of the aperture. This class allows an operator to control and scan
+    the position 'center' and opening 'width' of the aperture. The 'JJSlits'
+    class instantiates four sub-devices: 'JJSlits.xwidth', 'JJSlits.xcenter',
+    'JJSlits.ywidth', 'JJSlits.ycenter'.
```
flake8 is still showing a line length violation here, so let's adjust slightly
```suggestion
    dimension, one motor defines the aperture opening, and another determines
    the position of the aperture. This class allows an operator to control
    and scan the position 'center' and opening 'width' of the aperture. The
    'JJSlits' class instantiates four sub-devices: 'JJSlits.xwidth',
    'JJSlits.xcenter', 'JJSlits.ywidth', 'JJSlits.ycenter'.
```
